### ISSUE TITLE
index grow very large and shrink only after restart

### ISSUE THREAD

**Issue by andrewufrank** [syncthing/syncthing#3245]
i have 3 boxes with odroid armhf debian jessie 13.4. running syncthing continuously 24/7  (and nothing else) with each about 300 GB of data in sync. they crash occasionally (say every 2 weeks) because the index file grow to use all space on the partition. after restart (from systemctl) the index is compacted. 
example: the partition is 11GB and the index is filling this. after restart, the partition has 6.4 GB available - thus the index shrank by 6 GB. 

i think a workaround is to restart syncthing once per week (using a cronjob). is this a good idea?
what documentation would help you to track this issue? 

thank you for a wonderful tool!

**Comment by calmh** [comment]
The index keeps data for all connected devices. It growing as things connect and then shrinking on restart (when it only has data for itself) is thus expected.

**Comment by andrewufrank** [comment]
the problem here is that syncthing is running unattended and not restarted for long periods (say weeks) and the index seems to grow continously. could you automatically check if it is (say) doubled in size and then do the restart like operation to compact the index back to a reasonable size - or at least to trigger this action if the disk partition is nearly full. (not the data disk, the disk where the index is stored). 
thank you for the quick respnse!

**Comment by calmh** [comment]
The database does auto compaction and should not continue to grow on its own. If it does it's a bug in the database layer. Can you log or graph the database size over a longer period of time (days)?

**Comment by timwhite** [comment]
I'm experiencing instances where my DB is growing large as well, using all free disk space. This causes a crash then of course the database shrinks on restart. Since going to version 0.13 I am syncing more data, approximately 1.3TB, mostly between 2 devices but a bit with a few other devices.

After freeing more space on my home partition, and clicking "Override changes" (as the machine is a Master for a large 1TB folder), the database quickly shot up to 10.2GiB with ~6100 files in it. It appears to have stabilised there for a minute, before suddenly shrinking the whole thing down to 1.6GiB at 886 files. So it's much larger while scanning, but then compacted?

I understand that there is a "price" for syncing large amounts of data, but if after scanning the 10.2GiB becomes 1.6, is there something we can do to improve this during scanning? Perhaps detecting disk free so it can optimise when a homedir isn't as big as it needs?

**Comment by andrewufrank** [comment]
do you have a suggestion (cronjob? ) how i could log the size of the database over several days - i am not experienced with this kind of commands, but have not problem of adding a cronjob you suggest. 
thank you!
andrew

**Comment by calmh** [comment]
Sure. Cron something like this every five minutes or so:

`(date; du -s ~/.config/syncthing) >> ~/size.log`

**Comment by canton7** [comment]
(Beware that `~` may not be set in a cron job - it might be better to use an explicit path there (`/home/you/.config/syncthing` and `/home/you/size.log`).

**Comment by andrewufrank** [comment]
on another computer i just had the same problem (this time an intel cpu): the index was 17.1 GB. after the rebuild is seems to be less than 1 GB. 
i just added the cron job:

`5,10,15,20,25,30,35,40,45,50,55 *  * * * ; du -s /home/frank/.config/syncthing  >> /home/frank/synsize.log
`
but it does not produce a file `synsize.log`
what is wrong?

**Comment by andrewufrank** [comment]
i attach a file which shows the index to grow to 15 GB (for a total of 160GB of files in sync) and then shrinking to 0.5 GB. i observed similar behaviour on other nodes. i hope the index stays that small.

the entries are not the same amount of time apart, but triggered by hand. i hope it helps to find the bug!
[copy of synsize.txt](https://github.com/syncthing/syncthing/files/302880/copy.of.synsize.txt)

**Comment by andrewufrank** [comment]
i observed the size of the index and it oscillates between 0.4 GiB and 15 GiB as you can see in the following log of `du -s .config/syncthing`. it appears that whenever another node is restarted this node is building a large index (typeicaly 3..4 GiB) which is then discareded when the startup of the other node is completed. (there are approx 150,000 files for  a total of 160 GiB disk space on this computer. some of the other nodes hold 300,000 files for 400 GiB).

if more than one node starts at the same time, the index can grow to more than 7 GiB (and in some instances above 10GiB) to shrink afterwards - which seems unreasonable. syncthing may stop for lack of disk space and must be restarted manually (sometimes the index is even corrupted). 
i have now provided sufficient disk space for the index, to avoid the problem, but i think it is warranted to check the logic of the code to see why large index files emerge to be compacted. would it be possible to build one index file after the other and compact when one done?  

i hope the observations help and if you need more and better data, please give me precise instructions what i can do.

`-- automatic restarted synthing when upgrading from 13.4 to 13.5
420380	/home/frank/.config/syncthing
489812	/home/frank/.config/syncthing
896256	/home/frank/.config/syncthing
-- not explanation for shrinking, probably starting now to sync with other clients
459688	/home/frank/.config/syncthing
1629844	/home/frank/.config/syncthing
2559524	/home/frank/.config/syncthing
3556060	/home/frank/.config/syncthing
4390488	/home/frank/.config/syncthing
8045760	/home/frank/.config/syncthing
10366632	/home/frank/.config/syncthing
10366820	/home/frank/.config/syncthing
10366836	/home/frank/.config/syncthing
16370836	/home/frank/.config/syncthing
16451708	/home/frank/.config/syncthing
-- very rapid growth - less than a minute between 

16575616	/home/frank/.config/syncthing
16608344	/home/frank/.config/syncthing
-- stop and restart 

16608344	/home/frank/.config/syncthing
-- restart 
483504	/home/frank/.config/syncthing
419340	/home/frank/.config/syncthing
265992	/home/frank/.config/syncthing
265992	/home/frank/.config/syncthing
-- seems to sync with one or two clients -- they have 0.5 GB in .config/syncthing/index 
1059056	/home/frank/.config/syncthing
1392120	/home/frank/.config/syncthing
2374128	/home/frank/.config/syncthing
2666052	/home/frank/.config/syncthing
4522348	/home/frank/.config/syncthing
4995248	/home/frank/.config/syncthing
5760548	/home/frank/.config/syncthing
5964492	/home/frank/.config/syncthing
5981228	/home/frank/.config/syncthing
-- several hours of running - 
364464	/home/frank/.config/syncthing
363872	/home/frank/.config/syncthing
361816	/home/frank/.config/syncthing
379984	/home/frank/.config/syncthing
-- next day 

1243108	/home/frank/.config/syncthing
-- two hours, one additional client came up- the client reduced his index by 1.5 GB
-- and become inactive (no connections, not trying to open connections again 
-- after some read timeouts
376860	/home/frank/.config/syncthing
376860	/home/frank/.config/syncthing
-- restart other client 

376924	/home/frank/.config/syncthing
1205628	/home/frank/.config/syncthing
1218388	/home/frank/.config/syncthing
1285636	/home/frank/.config/syncthing
-- active sync with the new client, which has 300GB files, half shared with this node
4664644	/home/frank/.config/syncthing
4664856	/home/frank/.config/syncthing
-- several hours of processig the sync with the new node (not completed) 
459200	/home/frank/.config/syncthing
459336	/home/frank/.config/syncthing

462120	/home/frank/.config/syncthing
452612	/home/frank/.config/syncthing
-- restart another client which lost index (which was 5GB) and is rebuilding 
1180524	/home/frank/.config/syncthing
1180620	/home/frank/.config/syncthing
843412	/home/frank/.config/syncthing
407296	/home/frank/.config/syncthing
576876	/home/frank/.config/syncthing
3628640	/home/frank/.config/syncthing
3647508	/home/frank/.config/syncthing
4221908	/home/frank/.config/syncthing
4222920	/home/frank/.config/syncthing
5956664	/home/frank/.config/syncthing
876048	/home/frank/.config/syncthing
321256	/home/frank/.config/syncthing
304772	/home/frank/.config/syncthing
304772	/home/frank/.config/syncthing
-- restarting a client 
1540440	/home/frank/.config/syncthing
1956720	/home/frank/.config/syncthing
637560	/home/frank/.config/syncthing
3045304	/home/frank/.config/syncthing
3174172	/home/frank/.config/syncthing
399392	/home/frank/.config/syncthing
399440	/home/frank/.config/syncthing`

**Comment by calmh** [comment]
Is your device quite busy CPU wise when receiving the index data from the other devices?

To be clear this is all deep inside the database package that we are mostly just users of, but it seems like the compaction code in that package either isn't triggered or just plain doesn't get time to run between the updates that are streaming in.

**Comment by andrewufrank** [comment]
the main device (from which the `du -s`  measures are, is not really busy (it is my desktop); syncthing there uses occasionally more than a few % cpu, but it uses another resource (interrupt?) and causes unresponsiveness for other applications - i did not think that this might be related, but it seems possible. 
the other devices are odroid with debian and show during the sync up to 50% cpu (nothing else is running there). usually they around 5% cpu. 

what database are you using? i assume it stores metainformation for the files? is compaction run as a separate process and can be observed? 

at the moment this is not a problem for me as i realized that i can prepare the disk space for very large index occasionally. i hope i do not see crashes anymore.

thank you for the information!

**Comment by arneko** [comment]
Looks like a duplicate of https://github.com/syncthing/syncthing/issues/2400.  This is definitely an issue, also because it's hard to foresee just HOW large a database will grow. It's definitely not just startupSize*#devices.
Giving Syncthing lots of breathing room on large repos surely helps (moved over my ST-index to another partition via symlinks). And for some reason I never had this problem on my little ARM-server, whose OS used to live on a 16GiB SD-Card.

**Comment by AudriusButkevicius** [comment]
The DB size should be more or less stable with 0.14.2 and introduction of delta indexes.
